clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -454,64 +467,116 @@ private int append(Map<TopicPartition, List<ConsumerRecord<K, V>>> drained,
-            return client.send(node, ApiKeys.LIST_OFFSETS, request)
-                    .compose(new RequestFutureAdapter<ClientResponse, Long>() {
+    private RequestFuture<Map<TopicPartition, OffsetAndTimestamp>> sendListOffsetRequests(final Map<TopicPartition, Long> timestampsToSearch) {
+        for (TopicPartition topicPartition : timestampsToSearch.keySet()) {
+            PartitionInfo info = metadata.fetch().partition(topicPartition);
+            if (info == null) {
+                metadata.add(topicPartition.topic());
+                log.debug("Partition {} is unknown for fetching offset, wait for metadata refresh", topicPartition);
+                return RequestFuture.staleMetadata();
+            } else if (info.leader() == null) {
+                log.debug("Leader for partition {} unavailable for fetching offset, wait for metadata refresh", topicPartition);
+                return RequestFuture.leaderNotAvailable();
+            }
+        }
+
+        // Group the partitions by node.
+        final Map<Node, Map<TopicPartition, Long>> timestampsToSearchByNode = new HashMap<>();
+        for (Map.Entry<TopicPartition, Long> entry: timestampsToSearch.entrySet()) {
+            Node node = metadata.fetch().leaderFor(entry.getKey());
+            Map<TopicPartition, Long> topicData = timestampsToSearchByNode.get(node);
+            if (topicData == null) {
+                topicData = new HashMap<>();
+                timestampsToSearchByNode.put(node, topicData);
+            }
+            topicData.put(entry.getKey(), entry.getValue());
+        }
+
+        final RequestFuture<Map<TopicPartition, OffsetAndTimestamp>> listOffsetRequestsFuture = new RequestFuture<>();
+        final Map<TopicPartition, OffsetAndTimestamp> timestampOffsets = new HashMap<>();
+        for (Map.Entry<Node, Map<TopicPartition, Long>> entry : timestampsToSearchByNode.entrySet()) {
+            sendListOffsetRequest(entry.getKey(), entry.getValue())
+                    .addListener(new RequestFutureListener<Map<TopicPartition, OffsetAndTimestamp>>() {
+                        @Override
+                        public void onSuccess(Map<TopicPartition, OffsetAndTimestamp> value) {
+                            timestampOffsets.putAll(value);
+                            if (timestampOffsets.size() == timestampsToSearch.size() && !listOffsetRequestsFuture.isDone())
+                                listOffsetRequestsFuture.complete(timestampOffsets);
+                        }
+
                         @Override
-                        public void onSuccess(ClientResponse response, RequestFuture<Long> future) {
-                            handleListOffsetResponse(topicPartition, response, future);
+                        public void onFailure(RuntimeException e) {
+                            // This may cause all the requests to be retried, but should be rare.
+                            if (!listOffsetRequestsFuture.isDone())
```

Comment:
This "check and act" doesn't seem safe. Maybe we need a new method in `RequestFuture`:

```java
boolean raiseIfNotDone(RuntimeException e);
```